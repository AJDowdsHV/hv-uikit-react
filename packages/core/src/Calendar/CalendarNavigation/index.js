export { default as HvComposedNavigation } from "./ComposedNavigation";
export { default as HvNavigation } from "./Navigation";
export { default as HvMonthSelector } from "./MonthSelector";
