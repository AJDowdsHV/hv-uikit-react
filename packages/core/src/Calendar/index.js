export { default } from "./Calendar";
export { default as HvCalendarHeader } from "./CalendarHeader";
