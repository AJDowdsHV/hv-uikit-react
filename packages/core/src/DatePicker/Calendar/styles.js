const styles = theme => ({
  calendarWrapper: {
    overflow: "hidden",
    backgroundColor: `${theme.hv.palette.atmosphere.atmo1}`,
    padding: `${theme.hv.spacing.sm}px`
  },
  calendarGrid: {
    display: "flex",
    flexFlow: "wrap",
    width: "280px"
  },
  navigationContainer: {
    display: "flex",
    justifyContent: "space-between",
    padding: `${theme.hv.spacing.xs}px 0`
  },
  navigationMonth: {
    minWidth: "160px"
  },
  calendarDay: {
    display: "flex",
    justifyContent: "center",
    flexDirection: "column",
    textAlign: "center",
    height: "40px",
    width: "40px"
  },
  calendarDate: {
    display: "flex",
    justifyContent: "center",
    flexDirection: "column",
    textAlign: "center",
    height: "40px",
    width: "40px",
    "&:hover": {
      backgroundColor: `${theme.hv.palette.atmosphere.atmo4}`,
      cursor: "pointer"
    }
  },
  calendarDateNotInMonth: {
    color: `${theme.hv.palette.atmosphere.atmo7}`
  },
  calendarDateSelected: {
    backgroundColor: `${theme.hv.palette.accent.acce1}`,
    color: `${theme.hv.palette.atmosphere.atmo1}`,
    "&:hover": {
      backgroundColor: `${theme.hv.palette.accent.acce1}`,
      color: `${theme.hv.palette.atmosphere.atmo1}`
    }
  },
  calendarDateInvalid: {
    "&:hover": {
      cursor: "not-allowed",
      backgroundColor: "transparent"
    }
  },
  calendarMonthlyGrid: {
    top: "0",
    height: "calc(100% - 120px)",
    marginTop: "100px",
    marginLeft: "-20px",
    display: "flex",
    zIndex: "10",
    padding: "0 20px",
    position: "absolute",
    flexFlow: "wrap",
    alignContent: "center",
    justifyContent: "space-evenly",
    backgroundColor: "#FFFFFF"
  },
  normalWidth: {
    width: "calc(100% - 2px)" // 2px for the borders.
  },
  rangeModeWidth: {
    width: "calc((100% / 2) - 2px)" // 100% divided by two to handle the range situation minus 2px for the borders.
  },
  calendarMonthlyCell: {
    display: "flex",
    justifyContent: "center",
    flexDirection: "column",
    textAlign: "center",
    height: "40px",
    width: "92px",
    "&:hover": {
      backgroundColor: `${theme.hv.palette.atmosphere.atmo4}`,
      cursor: "pointer"
    }
  },
  calendarMonthlyCellSelected: {
    backgroundColor: `${theme.hv.palette.accent.acce1}`,
    color: `${theme.hv.palette.atmosphere.atmo1}`,
    "&:hover": {
      backgroundColor: `${theme.hv.palette.accent.acce1}`,
      color: `${theme.hv.palette.atmosphere.atmo1}`
    }
  }
});

export default styles;
