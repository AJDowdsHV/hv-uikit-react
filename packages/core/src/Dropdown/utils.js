/**
 * Filter selected elements.
 *
 * @param {Object} list - the list to filter
 * @returns {Array} - the selected elements
 */
const getSelected = list => (list ? list.filter(elem => elem.selected) : []);

/**
 * Gets the selection label according to selection.
 *
 * @param {Object} list - the list to filter the selected elements from
 * @param {Object} labels - the labels to extract the textual values for the label
 * @param {Boolean} multiselect - if "true" the label will have a different format
 *
 * @returns {String} - the selection label
 */
const getSelectionLabel = (list, labels, multiSelect) => {
  const selected = getSelected(list);
  const hasSelection = selected.length > 0;
  const isSingleSelection = selected.length === 1;

  let selectionLabel = multiSelect ? labels.selectAll : labels.select;

  if (hasSelection && isSingleSelection) {
    selectionLabel = selected[0].label;
  } else if (hasSelection && multiSelect) {
    selectionLabel = `${labels.multiSelectionAction} ${selected.length} ${labels.multiSelectionConjunction} ${list.length}`;
  }

  return selectionLabel;
};

export { getSelectionLabel, getSelected };
