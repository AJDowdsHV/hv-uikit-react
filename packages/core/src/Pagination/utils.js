import { useCallback, useState } from "react";

export const setColor = (condition) => (condition ? "atmo5" : undefined);

export const getSafePage = (inPage, page, pages) =>
  Number.isNaN(inPage) ? page : Math.min(Math.max(inPage, 0), pages - 1);

export const usePageInput = (initialPage) => {
  const [page, setPage] = useState(initialPage + 1);

  const handleChange = useCallback((evt, newPage) => setPage(newPage), []);

  return [page, handleChange];
};
