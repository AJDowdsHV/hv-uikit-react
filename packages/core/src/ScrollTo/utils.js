export const getScrollTop = (c = window) => {
  if (c === window) {
    return (
      window.scrollY ||
      window.pageYOffset ||
      (document.documentElement && document.documentElement.scrollTop) ||
      document.body.scrollTop
    );
  }

  return c.scrollTop;
};

export const verticalScrollOffset = (t, c = window) => {
  if (c === window) {
    return (t?.getBoundingClientRect?.().top || 0) + (window.scrollY || window.pageYOffset);
  }
  if (getComputedStyle(c).position !== "static") {
    return t.offsetTop;
  }

  return t.offsetTop - c.offsetTop;
};

export const horizontalScrollOffset = (t, c = window) => {
  if (c === window) {
    return (t?.getBoundingClientRect?.().left || 0) + (window.scrollX || window.pageXOffset);
  }
  if (getComputedStyle(c).position !== "static") {
    return t.offsetLeft;
  }

  return t.offsetLeft - c.offsetLeft;
};

export const scrollElement = (element, container, offset = 0, direction) => {
  if (direction === "row") {
    const elemLeft = horizontalScrollOffset(element, container);
    container?.scrollTo?.({
      left: elemLeft - offset,
      behavior: "smooth",
    });
  } else {
    const elemTop = verticalScrollOffset(element, container);
    container?.scrollTo?.({
      top: elemTop - offset,
      behavior: "smooth",
    });
  }
  element.focus({ preventScroll: true });
};

export const isScrolledToTheBottom = (container) => {
  const containerScrollTop = getScrollTop(container);

  if (container === window) {
    // accounting for cases where html/body are set to height:100%
    const scrollHeight =
      (document.documentElement && document.documentElement.scrollHeight) ||
      document.body.scrollHeight;

    return containerScrollTop + window.innerHeight >= scrollHeight;
  }

  return containerScrollTop + container.offsetHeight >= container.scrollHeight;
};

export const findFirstVisibleElement = (container, options, offset) => {
  const boundsTop = verticalScrollOffset(container);

  let i = 0;
  // find index of first element whose top is still visible inside the container
  for (; i < options.length; i += 1) {
    const ele = document.getElementById(options[i].value);

    if (ele) {
      const elemTop = verticalScrollOffset(ele) - (options[i].offset || offset);

      if (elemTop > boundsTop) {
        break;
      }
    }
  }

  // return the previous index, the element that last scrolled past the top
  return i - 1;
};
