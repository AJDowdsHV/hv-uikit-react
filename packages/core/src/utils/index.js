export * from "./KeyboardUtils";

export { default as isBrowser } from "./browser";

export * from "./setId";

export * from "./focusableElementFinder";

export * from "./sizes";

export { default as useImageLoaded } from "./useImageLoaded";
