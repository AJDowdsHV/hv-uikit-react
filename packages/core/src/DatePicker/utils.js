import { REPRESENTATION_VALUES } from "../Calendar/enums";
import {
  DEFAULT_LOCALE,
  getFormattedDate,
  getMonthName,
  isDate,
  isSameMonth,
  isValidLocale,
} from "../Calendar/utils";

/**
 * Validates if a given date is visible according to the provided visible month and year numbers.
 *
 * @param {Date} date The date object to be validated.
 * @param {Object} visibleDate - The month and year numeric values of a visible date.
 * @param {number} visibleDate.visibleMonth number The number of the month (1 to 12).
 * @param {number} visibleDate.visibleYear number The number of the year.
 * @returns {boolean} `true` if the date is considered visible, `false` otherwise
 */
export const isVisibleDate = (date, visibleDate) => {
  const { visibleYear, visibleMonth } = visibleDate;
  const date1 = new Date(visibleYear, visibleMonth - 1, 1);
  return isSameMonth(date, date1);
};

export const validateDate = (date) => (isDate(date) && date) || new Date();

export const validateLocale = (locale) => (isValidLocale(locale) ? locale : DEFAULT_LOCALE);

export const getFormattedDateRange = (date, locale, rep = REPRESENTATION_VALUES.SHORT) => {
  const { startDate, endDate } = date;
  const monthYear = `${getMonthName(startDate, locale, rep)} ${startDate.getFullYear()}`;
  return `${startDate.getDate()} - ${endDate.getDate()} ${monthYear}`;
};

export const getSingleDateLabel = (date, locale) =>
  isDate(date) ? getFormattedDate(date, locale) : "";

export const getRangeDateLabel = ({ startDate, endDate }, locale) => {
  if (!(isDate(startDate) && isDate(endDate))) return getSingleDateLabel(startDate);

  return isSameMonth(startDate, endDate)
    ? getFormattedDateRange({ startDate, endDate }, locale)
    : `${getFormattedDate(startDate, locale)} - ${getFormattedDate(endDate, locale)}`;
};

export const getDateLabel = (date, rangeMode, locale) =>
  rangeMode ? getRangeDateLabel(date, locale) : getSingleDateLabel(date, locale);
