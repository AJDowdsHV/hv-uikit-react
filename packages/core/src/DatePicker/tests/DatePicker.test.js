import React from "react";
import { mount } from "enzyme";

import { Calendar as CalendarIcon } from "@hv/uikit-react-icons";
import { HvActionContainer, HvProvider, HvCalendar, HvDatePicker } from "../..";
import SingleCalendar from "../../Calendar/SingleCalendar";
import { isSameDay, makeUTCDate } from "../../Calendar/utils";

describe("[v3] <DatePicker /> with minimum configuration", () => {
  let wrapper;
  let DatePickerComponent;

  beforeEach(async () => {
    wrapper = mount(
      <HvProvider>
        <HvDatePicker />
      </HvProvider>
    );
    DatePickerComponent = wrapper.find(HvDatePicker);
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(HvDatePicker)).toMatchSnapshot();
  });

  it("should not render an actions component", () => {
    expect(DatePickerComponent.find(HvActionContainer).length).toBe(0);
  });

  it("should not render any Calendar component", () => {
    expect(wrapper.find(HvCalendar).length).toBe(0);
  });

  it("should add a <Calendar /> component to the container when clicking on the calendar icon", () => {
    DatePickerComponent.find(CalendarIcon).simulate("click");
    expect(wrapper.find(HvCalendar).length).toBe(1);
  });

  it("should open the calendar when clicking the calendar icon", () => {
    wrapper.find(CalendarIcon).simulate("click");

    expect(wrapper.find(HvCalendar).length).toBe(1);
  });

  it("should not show an Actions component when opening the calendar", () => {
    wrapper.find(CalendarIcon).simulate("click");

    expect(wrapper.find(HvActionContainer).length).toBe(0);
  });
});

describe("[v3] <DatePicker /> with Single Calendar mode", () => {
  let wrapper;

  beforeEach(async () => {
    wrapper = mount(
      <HvProvider>
        <HvDatePicker value={makeUTCDate(2019, 1, 1)} locale="en-US" />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(HvDatePicker)).toMatchSnapshot();
  });

  it("should not render any Calendar component", () => {
    expect(wrapper.find(HvCalendar).length).toBe(0);
  });

  it("should open the calendar when clicking the calendar icon", () => {
    wrapper.find(CalendarIcon).simulate("click");

    expect(wrapper.find(HvCalendar).length).toBe(1);
  });

  it("should not show an Actions component when opening the calendar", () => {
    wrapper.find(CalendarIcon).simulate("click");

    expect(wrapper.find(HvActionContainer).length).toBe(0);
  });

  it("should have the Calendar component with the same selected date as the received value property", () => {
    wrapper.find(CalendarIcon).simulate("click");
    const calendarInstance = wrapper.find(HvDatePicker);

    const sameDay = isSameDay(calendarInstance.prop("value"), makeUTCDate(2019, 1, 1, 12));
    expect(sameDay).toEqual(true);
  });
});

describe("[v3] <DatePicker /> with Range Calendar mode", () => {
  let wrapper;

  beforeEach(async () => {
    wrapper = mount(
      <HvProvider>
        <HvDatePicker
          rangeMode
          locale="en-US"
          startValue={makeUTCDate(2019, 1, 5, 12)}
          endValue={makeUTCDate(2019, 1, 10, 12)}
        />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should not render any Calendar component", () => {
    expect(wrapper.find(HvCalendar).length).toBe(0);
  });

  it("should open two calendars when clicking the calendar icon", () => {
    wrapper.find(CalendarIcon).simulate("click");

    expect(wrapper.find(SingleCalendar).length).toBe(2);
  });

  it("should show an Actions component when opening the calendars", () => {
    wrapper.find(CalendarIcon).simulate("click");

    expect(wrapper.find(HvActionContainer).length).toBe(1);
  });
});

describe("[v3] <DatePicker /> with custom properties", () => {
  let wrapper;
  const labels = {
    applyLabel: "OK",
    cancelLabel: "GO BACK",
    title: "TESTING LABEL",
    placeholder: "I'M THE PLACEHOLDER",
  };

  beforeEach(async () => {
    wrapper = mount(
      <HvProvider>
        <HvDatePicker
          locale="en-US"
          value={makeUTCDate(2019, 1, 5, 12)}
          labels={labels}
          horizontalPlacement="left"
          showActions
          id="testingDatePicker"
        />
      </HvProvider>
    );
  });

  it("should be defined", () => {
    expect(wrapper).toBeDefined();
  });

  it("should render correctly", () => {
    expect(wrapper.find(HvDatePicker)).toMatchSnapshot();
  });

  it("should show an Actions component when opening the calendar", () => {
    wrapper.find(CalendarIcon).simulate("click");
    expect(wrapper.find(HvActionContainer).length).toBe(1);
  });
});
