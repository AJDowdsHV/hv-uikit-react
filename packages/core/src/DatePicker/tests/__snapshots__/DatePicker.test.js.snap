// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HvDatepicker calendar custom properties should render correctly 1`] = `
<div>
  <div
    class="HvDatePicker-root HvFormElement-root"
    id="testingDatePicker"
    locale="en-US"
  >
    <div
      class="HvBaseDropdown-root HvDatePicker-dropdown"
    >
      <div
        aria-expanded="false"
        aria-haspopup="dialog"
        class="HvBaseDropdown-anchor"
        role="combobox"
        tabindex="-1"
      >
        <div
          class="HvBaseDropdown-header"
          role="textbox"
          tabindex="0"
        >
          <div
            class="HvBaseDropdown-selection"
          >
            <p
              class="HvTypography-root HvTypography-placeholderText"
            >
              I'M THE PLACEHOLDER
            </p>
          </div>
          <div
            class="HvDatePicker-icon HvIconBase-root HvIconBase-s"
            name="Calendar"
          >
            <svg
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M10.5 2V0h-1v2h-3V0h-1v2H0v14h16V2zM15 15H1V3h14zM3 6h2v2H3zm4 0h2v2H7zm4 0h2v2h-2zm-8 4h2v2H3zm4 0h2v2H7zm4 0h2v2h-2z"
                fill="#414141"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HvDatepicker ranged calendar should render correctly 1`] = `
<div>
  <div
    class="HvDatePicker-root HvFormElement-root"
    locale="en-US"
  >
    <div
      class="HvBaseDropdown-root HvDatePicker-dropdown"
    >
      <div
        aria-expanded="false"
        aria-haspopup="dialog"
        class="HvBaseDropdown-anchor"
        role="combobox"
        tabindex="-1"
      >
        <div
          class="HvBaseDropdown-header"
          role="textbox"
          tabindex="0"
        >
          <div
            class="HvBaseDropdown-selection"
          >
            <p
              class="HvTypography-root HvTypography-normalText"
            >
              5 - 10 Jan 2019
            </p>
          </div>
          <div
            class="HvDatePicker-icon HvIconBase-root HvIconBase-s"
            name="Calendar"
          >
            <svg
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M10.5 2V0h-1v2h-3V0h-1v2H0v14h16V2zM15 15H1V3h14zM3 6h2v2H3zm4 0h2v2H7zm4 0h2v2h-2zm-8 4h2v2H3zm4 0h2v2H7zm4 0h2v2h-2z"
                fill="#414141"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HvDatepicker sample snapshot testing Main 1`] = `
<div>
  <div
    class="HvDatePicker-root HvFormElement-root"
    id="DatePicker"
    locale="en-US"
  >
    <div
      class="HvBaseDropdown-root HvDatePicker-dropdown"
    >
      <div
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-label="Date"
        class="HvBaseDropdown-anchor"
        role="combobox"
        tabindex="-1"
      >
        <div
          aria-label="Date"
          class="HvBaseDropdown-header"
          role="textbox"
          tabindex="0"
        >
          <div
            class="HvBaseDropdown-selection"
          >
            <p
              class="HvTypography-root HvTypography-placeholderText"
            >
              Select date
            </p>
          </div>
          <div
            class="HvDatePicker-icon HvIconBase-root HvIconBase-s"
            name="Calendar"
          >
            <svg
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M10.5 2V0h-1v2h-3V0h-1v2H0v14h16V2zM15 15H1V3h14zM3 6h2v2H3zm4 0h2v2H7zm4 0h2v2h-2zm-8 4h2v2H3zm4 0h2v2H7zm4 0h2v2h-2z"
                fill="#414141"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HvDatepicker single calendar should render correctly 1`] = `
<div>
  <div
    class="HvDatePicker-root HvFormElement-root"
    locale="en-US"
  >
    <div
      class="HvBaseDropdown-root HvDatePicker-dropdown"
    >
      <div
        aria-expanded="false"
        aria-haspopup="dialog"
        class="HvBaseDropdown-anchor"
        role="combobox"
        tabindex="-1"
      >
        <div
          class="HvBaseDropdown-header"
          role="textbox"
          tabindex="0"
        >
          <div
            class="HvBaseDropdown-selection"
          >
            <p
              class="HvTypography-root HvTypography-normalText"
            >
              1 Jan 2019
            </p>
          </div>
          <div
            class="HvDatePicker-icon HvIconBase-root HvIconBase-s"
            name="Calendar"
          >
            <svg
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M10.5 2V0h-1v2h-3V0h-1v2H0v14h16V2zM15 15H1V3h14zM3 6h2v2H3zm4 0h2v2H7zm4 0h2v2h-2zm-8 4h2v2H3zm4 0h2v2H7zm4 0h2v2h-2z"
                fill="#414141"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`HvDatepicker with minimum configuration should render correctly 1`] = `
<div>
  <div
    class="HvDatePicker-root HvFormElement-root"
    locale="en-US"
  >
    <div
      class="HvBaseDropdown-root HvDatePicker-dropdown"
    >
      <div
        aria-expanded="false"
        aria-haspopup="dialog"
        class="HvBaseDropdown-anchor"
        role="combobox"
        tabindex="-1"
      >
        <div
          class="HvBaseDropdown-header"
          role="textbox"
          tabindex="0"
        >
          <div
            class="HvBaseDropdown-selection"
          >
            <p
              class="HvTypography-root HvTypography-placeholderText"
            />
          </div>
          <div
            class="HvDatePicker-icon HvIconBase-root HvIconBase-s"
            name="Calendar"
          >
            <svg
              focusable="false"
              height="16"
              viewBox="0 0 16 16"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                class="color0"
                d="M10.5 2V0h-1v2h-3V0h-1v2H0v14h16V2zM15 15H1V3h14zM3 6h2v2H3zm4 0h2v2H7zm4 0h2v2h-2zm-8 4h2v2H3zm4 0h2v2H7zm4 0h2v2h-2z"
                fill="#414141"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
